###  贝叶斯拼写检查器 ###

In [36]:
import re,collections
def works(text):
    return re.findall('[a-z]+',text.lower())

def train(features):
    model = collections.defaultdict(lambda: 1)
    for f in features:
        model[f] += 1
    return model
NWORDS = train(works(open('big.txt').read()))
alphabet = 'abcdefghijklmnopqrstuvwxyz'
def edits1(word):
    n = len(word)
    return set([word[0:i]+word[i+1:] for i in range(n)] +                     # deletion
               [word[0:i]+word[i+1]+word[i]+word[i+2:] for i in range(n-1)] + # transposition
               [word[0:i]+c+word[i+1:] for i in range(n) for c in alphabet] + # alteration
               [word[0:i]+c+word[i:] for i in range(n+1) for c in alphabet])  # insertion

def known_edits2(word):
    return set(e2 for e1 in edits1(word) for e2 in edits1(e1) if e2 in NWORDS)

def known(words):
    return set(w for w in words if w in NWORDS)

def correct(word):
    candidates = known([word]) or known(edits1(word)) or known_edits2(word) or [word]
    return max(candidates,key=lambda w: NWORDS[w])

In [37]:
#appl #appla #learw #tess #morw
correct('knon')

'know'

### 求解：argmaxc P(c|w) -> argmaxc P(w|c) P(c) / P(w) ###

* P(c), 文章中出现一个正确拼写词 c 的概率, 也就是说, 在英语文章中, c 出现的概率有多大
* P(w|c), 在用户想键入 c 的情况下敲成 w 的概率. 因为这个是代表用户会以多大的概率把 c 敲错成 w
* argmaxc, 用来枚举所有可能的 c 并且选取概率最大的

In [24]:
# 把语料中的单词全部抽取出来，转成小写，并且去除单词中的特殊符号
def words(text):
    return re.findall('[a-z]',text.lower())

def train(features):
    model = collections.defaultdict(lambda :1)
    for f in features:
        model[f] += 1
    return model

NWORDS = train(words(open('big.txt').read()))

要是遇到我们从来没有遇到的新闻怎么办，假如说一个词拼写完全正确，但是语料中没有包含这个词，从而这个词也永远不会出现在训练集中。于是，我们就要返回出现这个词的概率为0.这种情况不太妙，因为概率为0代表了这个事件绝对不可能发生，而在我们的概率模型中，我们期望用一个很小的概率来代表这种情况。lamda:1

In [25]:
NWORDS

defaultdict(<function __main__.train.<locals>.<lambda>()>,
            {'t': 460735,
             'h': 294682,
             'e': 633813,
             'p': 98911,
             'r': 309546,
             'o': 386861,
             'j': 6437,
             'c': 144965,
             'g': 96908,
             'u': 138733,
             'n': 369005,
             'b': 73162,
             'k': 32799,
             'f': 120871,
             'a': 408084,
             'd': 215699,
             'v': 52379,
             's': 334892,
             'l': 198645,
             'm': 127062,
             'y': 90482,
             'i': 365619,
             'w': 100832,
             'x': 9808,
             'z': 3797,
             'q': 4572})

###  编辑距离: ###
两个词之间的编辑距离定义为使用了几次插入（在词中插入一个单字母），删除（删除一个单字母），交换（交换相邻两个字母），替换（把一个字母换成另一个）的操作，从一个词变到另一个词。

In [44]:
# 返回所有与单词word编辑距离为1的集合
def edits1(word):
    n = len(word)
    return set([word[0:i]+word[i+1:] for i in range(n)] +                     # deletion删除
               [word[0:i]+word[i+1]+word[i]+word[i+2:] for i in range(n-1)] + # transposition交换 
               [word[0:i]+c+word[i+1:] for i in range(n) for c in alphabet] + # alteration修改替换
               [word[0:i]+c+word[i:] for i in range(n+1) for c in alphabet])  # insertion插入

In [45]:
len(edits1('something'))

494

与something编辑距离为1的单词达到了494个

In [31]:
#返回所有与单词 w 编辑距离为 2 的集合
#在这些编辑距离小于2的词中间, 只把那些正确的词作为候选词
def edits2(word):
    return set(e2 for e1 in edits1(word) for e2 in edits1(e1))

In [33]:
len(edits2('something'))

114324

与something编辑距离为2的单词居然达到了114324个

优化：在这些编辑距离小于2的词中间，只把那些正确的词作为候选词，只能返回3个单词：‘smoothing’, ‘something’ 和 ‘soothing’

正常来说把一个元音拼成另一个元音的概率要大于辅音（因为人常常把hello打成hallo这样）；把单词的第一个字母拼错的概率会相对小一些，等等。但是为了简单起见，选择了一个简单的方法：编辑距离为1的正确单词比编辑距离为2的单词优先级高，而编辑距离为0的正确单词优先级比编辑距离为1的高。

In [47]:
def known(words):
    return set(w for w in words if w in NWORDS)

#如果known(set)非空, candidate 就会选取这个集合, 而不继续计算后面的
def correct(word):
    candidates = known([word]) or known(edits1(word)) or known(edits2(word)) or word
    print(candidates)
    return max(candidates,key=lambda w:NWORDS[w])

In [48]:
correct('knoq')

{'knox', 'knot', 'knob', 'know'}


'know'

In [53]:
NWORDS['know']

1049

In [51]:
NWORDS['knox']

3

In [52]:
NWORDS['knot']

17

In [54]:
NWORDS['knob']

3